File: 310971/Shabalinski/Task 310+320+330+340+350/BlogPlatform/publisher/src/main/java/com/task310/blogplatform/service/UserService.java
package com.task310.blogplatform.service;

import com.task310.blogplatform.dto.UserRequestTo;
import com.task310.blogplatform.dto.UserResponseTo;
import com.task310.blogplatform.exception.DuplicateException;
import com.task310.blogplatform.exception.EntityNotFoundException;
import com.task310.blogplatform.exception.ValidationException;
import com.task310.blogplatform.mapper.UserMapper;
import com.task310.blogplatform.model.User;
import com.task310.blogplatform.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Caching;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class UserService {
    private final UserRepository repository;
    private final UserMapper mapper;

    @Autowired
    public UserService(UserRepository repository, UserMapper mapper) {
        this.repository = repository;
        this.mapper = mapper;
    }

    @Caching(evict = {
        @CacheEvict(value = "users", allEntries = true)
    })
    public UserResponseTo create(UserRequestTo dto) {
        validateUserRequest(dto);
        
        // Check for duplicate login
        if (repository.findByLogin(dto.getLogin().trim()).isPresent()) {
            throw new DuplicateException("User with login '" + dto.getLogin() + "' already exists");
        }
        
        User user = mapper.toEntity(dto);
        User saved = repository.save(user);
        return mapper.toResponseDto(saved);
    }

    @Cacheable(value = "users", key = "'all'")
    public List<UserResponseTo> findAll() {
        return mapper.toResponseDtoList(repository.findAll());
    }

    @Cacheable(value = "users", key = "#id")
    public UserResponseTo findById(Long id) {
        if (id == null || id <= 0) {
            throw new ValidationException("Invalid user id");
        }
        return repository.findById(id)
                .map(mapper::toResponseDto)
                .orElseThrow(() -> new EntityNotFoundException("User not found with id: " + id));
    }

    @Caching(evict = {
        @CacheEvict(value = "users", key = "#id"),
        @CacheEvict(value = "users", key = "'all'")
    })
    public UserResponseTo update(Long id, UserRequestTo dto) {
        if (id == null || id <= 0) {
            throw new ValidationException("Invalid user id");
        }
        validateUserRequest(dto);
        User existing = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("User not found with id: " + id));
        
        // Check for duplicate login (excluding current user)
        repository.findByLogin(dto.getLogin().trim())
                .ifPresent(user -> {
                    if (!user.getId().equals(id)) {
                        throw new DuplicateException("User with login '" + dto.getLogin() + "' already exists");
                    }
                });
        
        mapper.updateEntityFromDto(dto, existing);
        User updated = repository.save(existing);
        return mapper.toResponseDto(updated);
    }

    @Caching(evict = {
        @CacheEvict(value = "users", key = "#id"),
        @CacheEvict(value = "users", key = "'all'")
    })
    public void delete(Long id) {
        if (id == null || id <= 0) {
            throw new ValidationException("Invalid user id");
        }
        if (!repository.existsById(id)) {
            throw new EntityNotFoundException("User not found with id: " + id);
        }
        repository.deleteById(id);
    }

    private void validateUserRequest(UserRequestTo dto) {
        if (dto == null) {
            throw new ValidationException("User data is required");
        }
        if (dto.getId() != null) {
            throw new ValidationException("Id must not be provided in request body");
        }
        if (dto.getLogin() == null || dto.getLogin().trim().isEmpty()) {
            throw new ValidationException("Login is required");
        }
        if (dto.getLogin().trim().length() < 2) {
            throw new ValidationException("Login must be at least 2 characters long");
        }
        if (dto.getPassword() == null || dto.getPassword().trim().isEmpty()) {
            throw new ValidationException("Password is required");
        }
        if (dto.getPassword().trim().length() < 8) {
            throw new ValidationException("Password must be at least 8 characters long");
        }
        if (dto.getFirstname() == null || dto.getFirstname().trim().isEmpty()) {
            throw new ValidationException("Firstname is required");
        }
        if (dto.getFirstname().trim().length() < 2) {
            throw new ValidationException("Firstname must be at least 2 characters long");
        }
        if (dto.getLastname() == null || dto.getLastname().trim().isEmpty()) {
            throw new ValidationException("Lastname is required");
        }
        if (dto.getLastname().trim().length() < 2) {
            throw new ValidationException("Lastname must be at least 2 characters long");
        }
    }
}

File: 351003/Pulish/publisher/app/services/user_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from app.dto.user import UserRequestTo, UserResponseTo
from app.models.user import User
from app.models.mark import Mark
from app.core.exceptions import NotFoundException, AppException


class UserService:
    def __init__(self, db: Session):
        self.db = db

    def create(self, dto: UserRequestTo) -> UserResponseTo:
        user = User(login=dto.login, password=dto.password,
                    firstname=dto.firstname, lastname=dto.lastname)
        try:
            self.db.add(user)
            self.db.commit()
            self.db.refresh(user)
        except IntegrityError:
            self.db.rollback()
            raise AppException("Login already exists", 40301, 403)
        return self._to_response(user)

    def find_all(self):
        users = self.db.query(User).all()
        return [self._to_response(u) for u in users]

    def find_by_id(self, id: int):
        user = self.db.query(User).filter(User.id == id).first()
        if not user:
            raise NotFoundException("User not found", 40401)
        return self._to_response(user)

    def update(self, dto: UserRequestTo):
        user = self.db.query(User).filter(User.id == dto.id).first()
        if not user:
            raise NotFoundException("User not found", 40401)

        user.login = dto.login
        user.firstname = dto.firstname
        user.lastname = dto.lastname
        try:
            self.db.commit()
            self.db.refresh(user)
        except IntegrityError:
            self.db.rollback()
            raise AppException("Login already exists", 40301, 403)
        return self._to_response(user)

    def delete(self, id: int):
        user = self.db.query(User).filter(User.id == id).first()
        if not user:
            raise NotFoundException("User not found", 40401)

        marks_to_check = set()
        for t in user.topics:
            for m in t.marks:
                marks_to_check.add(m.id)

        self.db.delete(user)
        self.db.commit()

        for m_id in marks_to_check:
            mark_in_db = self.db.query(Mark).filter(Mark.id == m_id).first()
            if mark_in_db and len(mark_in_db.topics) == 0:
                self.db.delete(mark_in_db)
        self.db.commit()

    def _to_response(self, user: User) -> UserResponseTo:
        return UserResponseTo(id=user.id, login=user.login,
                              firstname=user.firstname,
                              lastname=user.lastname)
